### ISSUE TITLE
Module resolution: Package Entrypoints not correctly used for types / source

### ISSUE THREAD

**Issue by jorenbroekema** [microsoft/TypeScript#56734]
### Demo Repo

https://github.com/jorenbroekema/ts-entrypoints-bug

### Which of the following problems are you reporting?

The module specifier resolves, but to the wrong file

### Demonstrate the defect described above with a code sample.

```js
import { isFoo } from "foo";

const whatever = isFoo();

export { whatever };
```
It resolves to the "types" entrypoint, which is okay for getting an understanding of the "type", but then when compiling it should use the "value" which is coming from the "default" entrypoint.

### Run `tsc --showConfig` and paste its output here

```json
{
    "compilerOptions": {
        "target": "esnext",
        "module": "esnext",
        "moduleResolution": "bundler",
        "declaration": true,
        "emitDeclarationOnly": true,
        "strict": true,
        "noImplicitThis": true,
        "alwaysStrict": true,
        "noImplicitAny": true
    },
    "files": [
        "./foo-test.ts"
    ],
    "include": [
        "**/*.ts"
    ],
    "exclude": [
        "node_modules",
        "**/coverage/*"
    ]
}
```

### Run `tsc --traceResolution` and paste its output here

```
======== Resolving module 'foo' from '/home/joren/code/type-entrypoint/foo-test.ts'. ========
Explicitly specified module resolution kind: 'Bundler'.
Resolving in CJS mode with conditions 'import', 'types'.
Found 'package.json' at '/home/joren/code/type-entrypoint/package.json'.
Loading module 'foo' from 'node_modules' folder, target file types: TypeScript, JavaScript, Declaration, JSON.
Searching all ancestor node_modules directories for preferred extensions: TypeScript, Declaration.
Found 'package.json' at '/home/joren/code/type-entrypoint/node_modules/foo/package.json'.
Entering conditional exports.
Matched 'exports' condition 'types'.
Using 'exports' subpath '.' with target './types/foo.d.ts'.
File '/home/joren/code/type-entrypoint/node_modules/foo/types/foo.d.ts' exists - use it as a name resolution result.
Resolved under condition 'types'.
Exiting conditional exports.
Resolving real path for '/home/joren/code/type-entrypoint/node_modules/foo/types/foo.d.ts', result '/home/joren/code/type-entrypoint/node_modules/foo/types/foo.d.ts'.
======== Module name 'foo' was successfully resolved to '/home/joren/code/type-entrypoint/node_modules/foo/types/foo.d.ts' with Package ID 'foo/types/foo.d.ts@1.0.0'. ========
```

Again, this looks fine, but I want to get the conditional export "default" for the actual value rather than the type, to prevent `'isFoo' only refers to a type, but is being used as a value here.ts(2693)`

### Paste the `package.json` of the *importing* module, if it exists

```json
{
  "name": "foo",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "keywords": [],
  "author": "",
  "license": "ISC",
  "exports": {
    ".": {
      "types": "./types/foo.d.ts",
      "default": "./index.js"
    }
  }
}
```

### Paste the `package.json` of the *target* module, if it exists

```json
{
  "name": "type-entrypoint",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "devDependencies": {
    "typescript": "^5.3.3"
  },
  "scripts": {
    "build": "tsc"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}
```

### Any other comments can go here

Perhaps I'm completely misunderstanding how conditional package entrypoints should behave for types vs values, feel free to educate me 😅

**Comment by fatcerberus** [comment]
The file specified by `types` is supposed to describe 1:1 the exports of the corresponding `.js` file.  In this case it doesn’t - the `.d.ts` exports a type alias, while the `.js` file exports a `const`.

**Comment by jorenbroekema** [comment]
> The file specified by `types` is supposed to describe 1:1 the exports of the corresponding `.js` file. In this case it doesn’t - the `.d.ts` exports a type alias, while the `.js` file exports a `const`.

Would there be a way to have the .d.ts file export a const as well then rather than a type alias? Or would that then break it again for type imports because they're now changed to be values?

**Comment by RyanCavanaugh** [comment]
You can import a dual-meaning type/value binding with `import type`. `import type` only controls whether or not you can use the import as a value in the importing file.

**Comment by jorenbroekema** [comment]
> You can import a dual-meaning type/value binding with `import type`. `import type` only controls whether or not you can use the import as a value in the importing file.

That explains how to force it into using the type, but that's what it imports by default already, how do I force it to import the value instead?

Can you give an example in the context of:

`node_modules/types/foo.d.ts`:
```ts
interface Options {}

export type isFoo = (opts: Options) => boolean;
```

Importing it in `foo-test.ts`:
```js
import { isFoo } from "foo"; // imports the type, but I want the value

const whatever = foo();

export { whatever };
```

**Comment by elizabeth-dev** [comment]
Hi

Isn't the issue just needing to change `/node_modules/foo/types/foo.d.ts` content
to:

```
interface Options {}

declare const isFoo: (opts: Options) => Boolean;
```

**Comment by jorenbroekema** [comment]
> Hi
> 
> I think the issue is just needing to change `/node_modules/foo/types/foo.d.ts` content to:
> 
> ```
> interface Options {}
> 
> declare const isFoo: (opts: Options) => Boolean;
> ```

Thanks, that solved it. so to summarize:
- `const isFoo = ...` means it's a value
- `type isFoo = ...` means it's a type (exclusively)
- `declare const isFoo: ...` means this is the type declaration that belongs to the value isFoo

I also see now that I made the mistake of not differentiating properly:
```ts
const isFoo = (opts: Options) => boolean; // error, `boolean` is type instead of value
```
and
```ts
const isFoo: (opts: Options) => boolean; // type
```

**Comment by fatcerberus** [comment]
Note: Don't use `Boolean`, always use `boolean` (same goes for `number` and `string`).  The former refers to "boxed" object values that you get by calling `new Boolean`.

**Comment by elizabeth-dev** [comment]
> Note: Don't use `Boolean`, always use `boolean` (same goes for `number` and `string`).  The former refers to "boxed" object values that you get by calling e.g. `new Boolean(true)`.

you are right......I wrote that from my phone and the autocorrect did their thing 🙈
